augur/db/import_/sqlite3.py

```diff
@@ -0,0 +1,92 @@
+import re
+import sqlite3
+
+from augur.db.defaults import DEFAULT_SEQUENCES_SEQ_NAME
+from augur.errors import AugurError
+from augur.io.file import StrainBasedFile
+from augur.io.metadata import Metadata
+from augur.io.sequences import Sequences
+from augur.io.sqlite3 import Sqlite3Database
+
+
+def import_(metadata_file:str, metadata_table_name:str,
+        sequences_file:str, sequences_table_name:str,
+        db_file:str, skip_strain_indexing:bool):
+    """Import data into a SQLite3 database file."""
+    database = Sqlite3Database(db_file)
+
+    if bool(metadata_file and metadata_table_name):
+        metadata = Metadata(metadata_file)
+        import_metadata(metadata, metadata_table_name, database)
+        if not skip_strain_indexing:
+            create_index(metadata, metadata_table_name, database)
+
+    if sequences_file and sequences_table_name:
+        # Get ID name from metadata if available.
+        id_name = metadata.id_name if bool(metadata_file and metadata_table_name) else None
+
+        sequences = Sequences(sequences_file, id_name)
+        import_sequences(sequences, sequences_table_name, database)
+        if not skip_strain_indexing:
+            create_index(sequences, sequences_table_name, database)
+
+
+def import_metadata(metadata:Metadata, table_name:str, database:Sqlite3Database):
+    """Import metadata into a SQLite3 database."""
+    try:
+        database.create_table(table_name, metadata.names)
+    except sqlite3.OperationalError as e:
+        database.delete()
+        raise AugurError(f'Failed to create table: {e}')
+
+    insert_statement = f"""
+        INSERT INTO {table_name}
+        VALUES ({','.join(['?' for _ in metadata.names])})
+    """
+    rows = metadata.rows()
+    try:
+        database.executemany(insert_statement, rows)
+    except sqlite3.ProgrammingError as e:
+        database.delete()
+
+        # Translate cryptic errors into meaningful messages.
+
+        column_count_error = re.match(r"Incorrect number of bindings supplied\. The current statement uses ([0-9]+), and there are ([0-9]+) supplied\.", str(e))
+        if column_count_error:
+            expected, observed = column_count_error.groups()
+            # TODO: detect and output the bad row for easier user debugging
+            raise AugurError(f'Failed to import {metadata.file}. A row has {observed} column(s) when {expected} are expected.')
+
+        raise AugurError(f'Failed to import {metadata.file}.')
+
+
+def import_sequences(sequences:Sequences, table_name:str, database:Sqlite3Database):
+    columns = [sequences.id_name, DEFAULT_SEQUENCES_SEQ_NAME]
+    try:
+        database.create_table(table_name, columns)
```

Comment:
Hmm. Another bigger picture design decision: I think we should be lightly standardizing the database schema. In particular, the names of the tables and their id columns. So the metadata id column should be static and the same as the sequences id column.

Without any standardization of schema, the SQLite database acts only as a container (albeit a useful one). With some light standardization, it can also be an interface that eases semantics-preserving interchange.

Reply:
How would you think to further standardize the schema? `augur db import` provides default values of `metadata` and `sequences` for the table name arguments and we require metadata to have a `strain` or `name` column through the `read_metadata` function. If we go with the single-table implementation you mention above, that would resolve the potential id column name mismatch.